mindmeld/active_learning/heuristics.py

```diff
@@ -0,0 +1,504 @@
+import numpy as np
+
+from ..constants import (
+    ENTROPY_LOG_BASE,
+    ACTIVE_LEARNING_RANDOM_SEED,
+)
+
+
+def stratified_random_sample(labels: List) -> List[int]:
+    """Reorders indices in evenly repeating pattern for as long as possible and then
+    shuffles and appends the remaining labels. The first part of this list will maintain
+    a uniform distrubition across labels, however, since the labels may not be perfectly
+    balanced the remaining portion will have a similar distribution as the original data.
+
+                 |-------- Evenly Repeating --------||--- Shuffled Remaining ----|
+    For Example: ["R","B","C","R","B","C","R","B","C","B","R","R","B","B","B","R"]
+
+    Args:
+        labels (List[str or int]): A list of labels. (Eg: labels = ["R", "B", "B", "C"])
+    Returns:
+        ranked_indices (List[int]): Indices corresponding to elements ranked by the heuristic.
+    """
+    np.random.seed(ACTIVE_LEARNING_RANDOM_SEED)
+
+    label_to_indices = _get_labels_to_indices(labels)
+    lowest_label_freq = min([len(indices) for indices in label_to_indices.values()])
+    avg_label_freq = len(labels) // len(label_to_indices)
+    sample_per_label = min(lowest_label_freq, avg_label_freq)
+
+    selected_indices = []
+    for i in range(sample_per_label):
+        for indices in label_to_indices.values():
+            selected_indices.append(indices[i])
+
+    remaining_indices = [i for i in range(len(labels)) if i not in selected_indices]
+    np.random.shuffle(remaining_indices)
+
+    ranked_indices = selected_indices + remaining_indices
+    return ranked_indices
+
+
+def _get_labels_to_indices(labels: List) -> defaultdict:
+    """Get a Dict mapping unique labels to indices in original data.
+    For Example: {"R": [1,3,5], "B": [2,6], "C":[4]}
+
+    Args:
+        labels (List[str or int]): A list of labels. (Eg: labels = ["R", "B", "B", "C"])
+    Returns:
+        labels_to_indices (defaultdict): Dictionary that maps unique labels to indices
+            where the label occurred in the original data.
+    """
+    labels_to_indices = defaultdict(list)
+    for i, label in enumerate(labels):
+        labels_to_indices[label].append(i)
+    for v in labels_to_indices.values():
+        np.random.shuffle(v)
+    return labels_to_indices
+
+
+def _convert_to_sample_ranks(ordered_sample_indices: List[int]):
+    """
+    Args:
+        ordered_sample_indices (List[int]): List of indices corresponding to values ordered
+        from least to greatest.
+    Returns:
+        sample_ranks (List[int]): List where the value at each index is the rank of the
+            corresponding sample.
+    """
+    sample_ranks = np.zeros(len(ordered_sample_indices), dtype=int)
+    for rank, sample_index in enumerate(ordered_sample_indices):
+        sample_ranks[sample_index] = rank
+    return sample_ranks
+
+
+def _ordered_indices_list_to_final_rank(ordered_sample_indices_list: List[List[int]]):
```

Comment:
this function should be a static method of the heuristic class since only inherited classes use this